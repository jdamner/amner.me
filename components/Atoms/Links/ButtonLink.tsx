import React from "react";

/* Components */
import Link, { type LinkProps } from "next/link";
import { ButtonClasses } from "./Button";

export default function ButtonLink(
  props: Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, keyof LinkProps> &
    LinkProps & {
      children?: React.ReactNode;
    } & React.RefAttributes<HTMLAnchorElement>,
) {
  return <Link className={ButtonClasses.join(" ")} {...props} />;
}
