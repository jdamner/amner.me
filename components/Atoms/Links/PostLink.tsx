"use client";

import React from "react";

/* Components */
import Link from "next/link";
import Image from "next/image";

/* Types */
import type { PostLinkType } from "../../../types";

/**
 * Post Link
 *
 * @param {Post} post
 * @return JSX.Element
 */
export default function PostLink({
  thumbnail,
  title,
  slug,
}: PostLinkType): React.JSX.Element {
  const classes = [
    "group",
    "block",
    "bg-white",
    "dark:bg-slate-900",
    "border-2",
    "border-slate-500",
  ];

  const linkClasses = [
    "bg-slate-900",
    "group-hover:bg-slate-700",
    "px-2",
    "text-lg",
    "leading-loose",
    "text-center text-slate-100",
  ];

  return (
    <Link href={slug} className={classes.join(" ")}>
      <Image
        src={thumbnail}
        alt={""}
        width={375}
        height={375}
        className="w-full brightness-90 dark:brightness:75 group-hover:brightness-105"
      />
      <h2 className={linkClasses.join(" ")}>{title}</h2>
    </Link>
  );
}
