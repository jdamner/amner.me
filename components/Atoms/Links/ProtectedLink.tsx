import React from "react";
import Link, { LinkProps } from "next/link";

export default function ProtectedLink(
  props: LinkProps &
    React.DetailedHTMLProps<
      React.HTMLAttributes<HTMLAnchorElement>,
      HTMLAnchorElement
    >,
) {
  // extract the HREF from the props, and make it
  // an on-click event... Stops bots from crawling
  // the link, but allows users to click it.
  const { href, ...rest } = props;

  const onClick = () => {
    window.location.href = href.toString();
  };

  return <Link {...rest} onClick={onClick} href="#" />;
}
