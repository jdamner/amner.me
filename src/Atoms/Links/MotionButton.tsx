'use client'
import React from "react";
/* API */
import { motion } from "framer-motion"
/* Types */
import type { MotionProps } from "framer-motion"

export default function MotionButton(props: React.DetailedHTMLProps<React.HTMLAttributes<HTMLButtonElement>, HTMLButtonElement> & { delay?: number }) {

    const letters = props.children?.toString().split('').map(
        (letter) => letter.replace(/\s/g, '\u00A0')
    ) ?? []

    return (
        <button {...props}>
            {letters.map((letter, index) => {
                const delay = props.delay + index * 0.05
                const motionProps: MotionProps = {
                    whileInView: {
                        translateY: [0, -3, 0],
                        transition: {
                            duration: 0.45,
                            delay,
                        },
                    },
                }
                return <motion.span {...motionProps} className="group-hover:underline inline-block" key={index}>{letter}</motion.span>
            })}
        </button>
    )
}
