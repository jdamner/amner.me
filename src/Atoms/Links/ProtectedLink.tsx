import React from "react";
import Link, { LinkProps } from "next/link";

export default function ProtectedLink(props: LinkProps & React.DetailedHTMLProps<React.HTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>) {

    // extract the HREF from the props, and make it 
    // an on-click event... Stops bots from crawling
    // the link, but allows users to click it.
    let { href, ...rest } = props;

    let onClick = () => {
        window.location.href = href.toString();
    }

    return (
        <Link {...rest} onClick={onClick} href='#' />
    )
}