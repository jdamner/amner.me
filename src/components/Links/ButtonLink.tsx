import Link from "next/link";
import { ButtonClasses } from "./Button";
import { useState } from "react";
import Spinner from "../Global/Spinner";
import { useRouter } from "next/router";
import { event } from "../../api/Insights";

export default function ButtonLink(props): JSX.Element {

	const [isClicked, setIsClicked] = useState(false);
	const router = useRouter();

	const handleClick = function (e: React.MouseEvent<HTMLAnchorElement>) {
		event('click', { name: e.currentTarget.innerText });
		setIsClicked(true);
		router.events.on('routeChangeComplete', () => {
			setIsClicked(false);
		});
	}

	return (
		<Link className={ButtonClasses.join(' ')} {...props} onClick={handleClick}>
			{ isClicked ? 
				<Spinner className='text-center' /> : props.children }
		</Link>
	)
}