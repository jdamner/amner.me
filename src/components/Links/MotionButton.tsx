/* API */
import { motion } from "framer-motion"
/* Types */
import type { MotionProps } from "framer-motion"

/**
 * Motion Button
 * 
 * Button where each letter is animated
 * 
 * @param {JSX.IntrinsicElements['button']} props
 * @returns JSX.Element
 */
export default function MotionButton(props: JSX.IntrinsicElements['button']): JSX.Element {

    const letters = props.children.toString().split('') as string[]
    // replace spaces with non-breaking spaces
    letters.forEach((letter, index) => {
        if (letter === ' ') {
            letters[index] = '\u00A0'
        }
    })

    return (
        <button {...props}>
            {letters.map((letter, index) => {
                const delay = props.tabIndex * 0.5 + index * 0.05
                const motionProps: MotionProps = {
                    style: {
                        display: 'inline-block',
                    },
                    whileInView: {
                        translateY: [0, -3, 0],
                        transition: {
                            duration: 0.45,
                            delay,
                        },
                    },
            
                }

                return <motion.span {...motionProps} className="group-hover:underline" key={index}>{letter}</motion.span>
            })
            }
        </button>
    )
}
