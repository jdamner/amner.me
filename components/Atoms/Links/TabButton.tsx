import React from "react";
/* API */
import { useReducedMotion } from "framer-motion";

/* Components */
import MotionButton from "./MotionButton";

export default function TabButton(
  props: React.DetailedHTMLProps<
    React.HTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  > & { active?: boolean; delay?: number },
) {
  const { active, ...rest } = props;
  const classes = [
    "text-xl md:text-3xl font-bold mr-3 md:mr-0 group hover:underline",
    active
      ? "text-gray-900 dark:text-gray-300"
      : "text-gray-300 dark:text-gray-500",
  ];

  const newProps = {
    ...rest,
    className: `${rest.className} ${classes.join(" ")}`,
  };

  return useReducedMotion() ? (
    <button {...newProps} />
  ) : (
    <MotionButton {...newProps} />
  );
}
