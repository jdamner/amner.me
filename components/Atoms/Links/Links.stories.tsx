import type { Meta, StoryObj } from "@storybook/react";
import { faker } from "@faker-js/faker";

import ButtonComponent from "./Button";
import ButtonLink from "./ButtonLink";
import MotionButton from "./MotionButton";
import InlineLink from "./Inline";
import PostLink from "./PostLink";
import ProtectedLink from "./ProtectedLink";
import TabButton from "./TabButton";
import { generatePostLink } from "../../../utils/storybook";

export default {
  title: "Atoms/Links & Buttons",
  component: ButtonComponent,
} as Meta;

faker.seed(123);

export const Button: StoryObj<typeof ButtonComponent> = {
  args: {
    children: faker.lorem.words(2),
  },
};

export const LinkButton: StoryObj<typeof ButtonLink> = {
  render: ButtonLink,
  args: {
    children: faker.lorem.words(2),
    href: faker.internet.url(),
  },
};

export const Motion: StoryObj<typeof MotionButton> = {
  render: MotionButton,
  args: {
    children: faker.lorem.words(2),
    delay: faker.number.float(),
  },
};

export const Inline: StoryObj<typeof InlineLink> = {
  render: InlineLink,
  args: {
    children: faker.lorem.words(2),
    href: faker.internet.url(),
  },
};

export const Post: StoryObj<typeof PostLink> = {
  render: (args) => (
    <div style={{ maxWidth: "300px" }}>
      <PostLink {...args} />
    </div>
  ),
  args: generatePostLink(),
};

export const Protected: StoryObj<typeof ProtectedLink> = {
  render: ProtectedLink,
  args: {
    children: faker.lorem.words(2),
    href: faker.internet.url(),
  },
};

export const Tab: StoryObj<typeof TabButton> = {
  render: TabButton,
  args: {
    children: faker.lorem.words(2),
    active: false,
    delay: faker.number.float(),
  },
};

export const ActiveTab: StoryObj<typeof TabButton> = {
  render: TabButton,
  args: {
    children: faker.lorem.words(2),
    active: true,
    delay: faker.number.float(),
  },
};
