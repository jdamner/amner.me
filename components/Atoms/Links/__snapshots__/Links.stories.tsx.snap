// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms/Links & Buttons ActiveTab smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <button delay="0.5315513701643795"
          class="undefined text-xl md:text-3xl font-bold mr-3 md:mr-0 group hover:underline text-gray-900 dark:text-gray-200"
  >
    atque agnosco
  </button>
</div>
`;

exports[`Atoms/Links & Buttons Button smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <button class="undefined border-2 border-slate-500 uppercase font-bold bg-white dark:bg-slate-900 px-4 py-2 text-slate-800 dark:text-slate-200 hover:bg-slate-100 dark:hover:bg-slate-800 hover:text-black dark:hover:text-white">
    subito summisse
  </button>
</div>
`;

exports[`Atoms/Links & Buttons Inline smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <a class="underline hover:text-slate-800 hover:no-underline"
     href="https://red-mountain.info/"
  >
    termes vomer
  </a>
</div>
`;

exports[`Atoms/Links & Buttons LinkButton smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <a class="border-2 border-slate-500 uppercase font-bold bg-white dark:bg-slate-900 px-4 py-2 text-slate-800 dark:text-slate-200 hover:bg-slate-100 dark:hover:bg-slate-800 hover:text-black dark:hover:text-white"
     href="https://remote-media.net/"
  >
    circumvenio curto
  </a>
</div>
`;

exports[`Atoms/Links & Buttons Motion smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <button delay="0.42310646059922874">
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      s
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      u
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      p
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      e
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      r
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      &nbsp;
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      d
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      e
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      p
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      r
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      e
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      c
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      a
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      t
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      o
    </span>
    <span class="group-hover:underline inline-block"
          style="transform: none;"
    >
      r
    </span>
  </button>
</div>
`;

exports[`Atoms/Links & Buttons Post smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <div style="max-width: 300px;">
    <a class="group flex flex-col-reverse md:flex-row bg-slate-200 p-5 border-4 border-slate-200 dark:border-slate-900 dark:bg-slate-900 hover:bg-slate-100 dark:hover:bg-slate-800 hover:border-slate-400 transition-all duration-100 ease-in-out"
       href="/suspendo-commodo-decens"
    >
      <div class="w-full text-left  text-2xl font-black">
        apud cruciamentum cui
        <div class="bg-slate-800 dark:bg-slate-300 h-1 w-full"
             style="margin-left: auto; max-width: 100%;"
        >
        </div>
      </div>
      <img alt
           loading="lazy"
           width="320"
           height="160"
           decoding="async"
           data-nimg="1"
           class="border-4 h-30 w-80 object-cover border-slate-800 dark:border-slate-300 brightness-90 dark:brightness:75 group-hover:brightness-105 bg-white dark:bg-slate-900 transition-all duration-100 ease-in-out"
           srcset="https://loremflickr.com/640/480?lock=5650369250066432&amp;w=384&amp;q=75 1x, https://loremflickr.com/640/480?lock=5650369250066432&amp;w=640&amp;q=75 2x"
           src="https://loremflickr.com/640/480?lock=5650369250066432&amp;w=640&amp;q=75"
           style="color: transparent;"
      >
    </a>
  </div>
</div>
`;

exports[`Atoms/Links & Buttons Protected smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <a href="/#">
    caries adulatio
  </a>
</div>
`;

exports[`Atoms/Links & Buttons Tab smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <button delay="0.44025717745535076"
          class="undefined text-xl md:text-3xl font-bold mr-3 md:mr-0 group hover:underline text-gray-600 dark:text-gray-400"
  >
    ratione auctor
  </button>
</div>
`;
